### PR TITLE
djbdns: fix crash due to data ulimit

### LINKED ISSUE #119066: djbdns' tinydns does not answer requests after a few requests
**Describe the bug**
According to multiple sources [0,1,2,3] (and my experience), tinydns does not work well with glibc-2.32.
The first queries are answered (hence the NixOS test pass!) but at some point none of them are answered.
The (cryptic) errors show that tinydns drops the requests [4].

A fix consists of increasing the "softlimit". It worked for tinydns (see the patch is below).

I'm not sure how we want to handle this. This will probably not be patched upstream, so we could ship the patchs in nixpkgs.
I tested increasing the softlimit from -d300000 to -d4500000 and it works. But there are other tools in the package which could be affected as well (grep softlimit in the source code to find them) . I don't use them, don't know how to use them and don't know if they are currently broken (hence I'm filling an issue, not a PR for now). I admit that I did not read the whole mail thread. Other may have find the solutions.

[0] http://cr-yp-to.996295.n3.nabble.com/bug-in-tinydns-djbdns-td23629.html
[1] https://marc.info/?l=qmail&m=160609563211809&w=2
[2] https://rjp.is/blogging/posts/2020/11/tinydns-snafu/
[3] http://cr-yp-to.996295.n3.nabble.com/on-openbsd-s-malloc-init-mmap-failed-td23189.html
[4] `ip:port:0000 / 0000` http://www.dqd.com/~mayoff/notes/djbdns/tinydns-log.html#noquery

 ```patch
diff --git a/tinydns-conf.c b/tinydns-conf.c
index d3a4ce5..8f43519 100644
--- a/tinydns-conf.c
+++ b/tinydns-conf.c
@@ -46,7 +46,7 @@ int main(int argc,char **argv)
 
   start("run");
   outs("#!/bin/sh\nexec 2>&1\nexec envuidgid "); outs(user);
-  outs(" envdir ./env softlimit -d300000 ");
+  outs(" envdir ./env softlimit -d4500000 ");
   outs(auto_home); outs("/bin/tinydns\n");
   finish();
   perm(0755);
```

**To Reproduce**
1. build a machine with the configuration below on commit 9e377a6ce42dccd9b624ae4ce8f978dc892ba0e2, for instance a VM: `nix-build -E '(import <nixpkgs/nixos> { configuration = ./configuration.nix; }).vm' -I nixpkgs=/path/to/nixpkgs`
2. start it, wait and log in
3. run multiple requests: `until ! dig bla.foo.bar @127.0.0.1 ; do echo ; done`
4. the server should stop answering after a few successful requests and its log shows that it gets the request and dropped them.

Log:
```
(...)
Apr 10 17:23:54 nixos tinydns-start[716]: 7f000001:dd8b:5a27 + 0001 bla.foo.bar
Apr 10 17:23:54 nixos tinydns-start[716]: 7f000001:a030:7172 + 0001 bla.foo.bar
Apr 10 17:23:54 nixos tinydns-start[716]: 7f000001:aec4:ccdf - 0001 bla.foo.bar
Apr 10 17:23:54 nixos kernel: mmap: tinydns (716): VmData 307200 exceed data ulimit 300000. Update limits or use boot option ignore_rlimit_data.
Apr 10 17:23:59 nixos tinydns-start[716]: 7f000001:aec4:0000 / 0000 .
Apr 10 17:24:04 nixos tinydns-start[716]: 7f000001:aec4:0000 / 0000 .
```
`configuration.nix`:
```nix
{ pkgs, ... }:
{
  services.tinydns = {
    enable = true;
    data = ''
      .foo.bar:1.2.3.4
      +.bla.foo.bar:1.2.3.4:300
    '';
  };
  environment.systemPackages = [ pkgs.bind.dnsutils ];
```
**Expected behavior**
The server should answer to all correct requests.

**Notify maintainers**
@basvandijk @jerith666

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.27, NixOS, 21.05pre-git (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.3774.a6847cb5460"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Note: I did not use the channel but a nixpkgs checkout on nixos-unstable (9e377a6ce42dccd9b624ae4ce8f978dc892ba0e2).

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by basvandijk**
###### Motivation for this change

This fixes: https://github.com/NixOS/nixpkgs/issues/119066.

When `djbdns` is repeatedly queried it will crash with:

```
mmap: tinydns (842): VmData 331776 exceed data ulimit 300000. Update limits or use boot option ignore_rlimit_data.
```

So we increase the softlimit from 300000 to 4500000 as suggested in
the issue.

The `tinydns` NixOS test has been extended with testing for the faulty
behaviour.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by basvandijk** [comment]
@GrahamcOfBorg test tinydns

**Review by ThibautMarty — CHANGES_REQUESTED** [review]
This combines my patch in #119066 and a test equivalent to #123131 so I'm fine with it.
I'd prefer to run more requests as the problem is related to hard-to-predict memory issues (I used 100, an arbitrary number larger than needed (bug triggered around the 10th iteration) with a acceptable runtime (4-5 seconds here).

Could you also add the link between the package and the test like https://github.com/NixOS/nixpkgs/pull/123131/commits/fa09d5d4d43b164fdc0e6f4426c28c14683cc3e2?

**Comment by basvandijk** [comment]
@ThibautMarty  good points. Done.

@GrahamcOfBorg test tinydns

**Review by jerith666 — APPROVED** [review]
Looks great to me; I agree upping the iteration count to 100 is a good change.  Thanks for pulling this together!

**Review by ThibautMarty — APPROVED** [review]

File: pkgs/tools/networking/djbdns/softlimit.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
++++ b/tinydns-conf.c	2021-11-08 17:23:06.181385437 +0000
+@@ -46,7 +46,7 @@
+
+   start("run");
+   outs("#!/bin/sh\nexec 2>&1\nexec envuidgid "); outs(user);
+-  outs(" envdir ./env softlimit -d300000 ");
++  outs(" envdir ./env softlimit -d4500000 ");
+   outs(auto_home); outs("/bin/tinydns\n");
+   finish();
+   perm(0755);
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/djbdns/softlimit.patch:12]: Please add a final new line that git can generate proper diff's.

Comment (basvandijk) [reply pkgs/tools/networking/djbdns/softlimit.patch]: Done. Thanks.

**Pull request merged**